compiler/rustc_expand/src/expand.rs

```diff
@@ -427,8 +427,10 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         let mut undetermined_invocations = Vec::new();
         let (mut progress, mut force) = (false, !self.monotonic);
         loop {
-            let Some((invoc, ext)) = invocations.pop() else {
+            if progress {
                 self.resolve_imports();
+            }
```

Comment:
Could you move this back into the `let Some((invoc, ext)) = invocations.pop() else { ... }` condition?
We don't need to resolve more imports if we still have invocations in the queue to process.

Reply:
Perhaps not, it is intended to ensure that the maximum number of resolved macros are available in the loop

Reply:
This is the reason for the existence of [multiple nested test](https://github.com/rust-lang/rust/pull/108729/files#diff-0d98fde7f8b0195e3ee9a54a83ee001ba6c1619d9314ff5ceae7b8fcad647cd3R39-R67) cases

Reply:
So, the purpose of running `resolve_imports` multiple times is to make it skip steps like step3, stpe5

Reply:
Or we can solve this problem by changing the value of `progress`

Reply:
>4. expand `bar!{}` but `resolved_macro_invocations` return `Err`

This part looks like a wrong behavior.

The crate at this point looks like
```rust
macro_rules! foo {
    () => {};
}

use foo as bar;

use bar as baz;

baz! {}
```
(which compiles ok by itself - https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=40e0c18cbb94afeb4a0cf11efd7d112c), and both `use foo as bar;` and `use bar as baz;` should be successfully resolved because we have just run `resolve_imports` to exhaustion.

Why does `resolved_macro_invocations` returns indeterminacy in this case?

Reply:
it caused by [lazy_resolutions](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/lib.rs#L1603), which in your example `baz` corresponds to ` resulutions.binding` is `Some(xxx)` so it returns success in  [check_usable](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L938-L942) returned successfully.

But in #97534, `baz` corresponds to `resulutions.binding` which is `None`, and it skips this step down and eventually returns `Err(Undetermined)`

Reply:
I suspect that the true root cause of the issue lies somewhere in this area.
There should be no difference in resolution results between https://github.com/rust-lang/rust/pull/108729#discussion_r1131657710 and https://github.com/rust-lang/rust/issues/97534.